feat: Recursion in public kernels and rollup circuits

PhilWindle opened the pull request:
This PR introduces recursion throughout the server side protocol circuits.

1. Public kernel circuits are now recursive.
2. Rollup circuits are now recursive.
3. Private kernel proofs are recursively verified in first server protocol circuit.
4. Some cleanup of old `Aggregation Object` code.
5. An e2e test and benchmarks for full transaction proving.
Closes #5241 #5245 #5984 #5985 #5986 #5989 #5990 #5991

PhilWindle committed:
Update serialisation

PhilWindle committed:
Merge branch 'master' into pw/serialise-public-inputs

PhilWindle committed:
Fixes

PhilWindle committed:
WIP

PhilWindle committed:
More serialisation

PhilWindle committed:
Fix

PhilWindle committed:
Updates constants

PhilWindle committed:
Merge branch 'master' into pw/serialise-more-inputs

PhilWindle committed:
WIP

PhilWindle committed:
WIP

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Recursion in all remaining protocol circuits

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Convert client proving test to full proving test

PhilWindle committed:
Increased timeout

PhilWindle committed:
Update earth file

PhilWindle committed:
Merge branch 'master' into pw/proving-e2e

PhilWindle committed:
File rename

PhilWindle committed:
Fixed typo

PhilWindle committed:
Only build 1 block

PhilWindle committed:
Merge branch 'master' into pw/proving-e2e

PhilWindle committed:
Test name

PhilWindle committed:
Merge branch 'master' into pw/proving-e2e

PhilWindle committed:
Merge branch 'pw/proving-e2e' of github.com:AztecProtocol/aztec-packages into pw/proving-e2e

PhilWindle committed:
Increase timeout

PhilWindle committed:
Fixes

PhilWindle committed:
Removed prover test

PhilWindle committed:
Fix

PhilWindle committed:
Formatting

PhilWindle committed:
Merge branch 'pw/proving-e2e' into pw/more-recursion

PhilWindle committed:
WIP

PhilWindle committed:
Fixes

PhilWindle committed:
Fixes for simulated circuits

PhilWindle committed:
Formatting

PhilWindle committed:
WIP

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Formatting

PhilWindle committed:
Fixes

PhilWindle committed:
Fixes

PhilWindle committed:
Fixes

PhilWindle committed:
Fix

PhilWindle committed:
Increase runner timeout

PhilWindle committed:
WIP

PhilWindle committed:
WIP

PhilWindle committed:
Cut down prover benchmark

PhilWindle committed:
Cleanup

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
WIP

PhilWindle committed:
WIP

PhilWindle committed:
Formatting

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Updated bench test

PhilWindle committed:
Merge fix

PhilWindle committed:
Reverted unrequired code

PhilWindle committed:
Test timeout values

PhilWindle committed:
Formatting

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

just-mitch commented on yarn-project/prover-client/src/orchestrator/orchestrator.ts:
```diff
@@ -276,13 +283,21 @@ export class ProvingOrchestrator {
+      tx,
+      inputs,
+      treeSnapshots,
+      provingState.privateKernelVerificationKeys.privateKernelToPublicCircuit,
```
I guess it doesn't matter based on the comment above, but I'm confused why we use the `ToPublic` keys here, but prepare for the base rollup?

PhilWindle replied:
So, if there are public functions, then this is the key that will be provided to the first public kernel to verify the previous private kernel. The key that gets given to the base rollup in this case will be the `publicKernelTail` and this is done once that circuit has been proven.

If there no public functions then it uses the `priveKernelCircuit` key provided further up to `prepareBaseRollupInputs`.

just-mitch reviewed (approved):
Some minor nits and suggestions on possible future cleanup, but overall looks good to me! Can't wait to get functions working! 🔥

PhilWindle commented on yarn-project/end-to-end/src/benchmarks/bench_prover.test.ts:
```diff
@@ -176,54 +176,58 @@ describe('benchmarks/proving', () => {
-      (await getTestContractOnPXE(1)).methods.emit_unencrypted(43),
-      (await getTestContractOnPXE(2)).methods.create_l2_to_l1_message_public(45, 46, EthAddress.random()),
-      (await getTokenContract(3)).methods.transfer(schnorrWalletAddress.address, recipient.address, 1000, 0),
+      //(await getTestContractOnPXE(1)).methods.emit_unencrypted(43),
```
Yeah, once we have public functions working we should be able to exercise fee payment flows as well.

PhilWindle commented on yarn-project/noir-protocol-circuits-types/src/index.ts:
```diff
@@ -555,12 +577,56 @@ export function convertPublicTeardownRollupInputsToWitnessMap(inputs: PublicKern
+ * @param inputs - The public kernel inputs.
+ * @returns The witness map
+ */
+export function convertPublicSetupRollupInputsToWitnessMap(inputs: PublicKernelCircuitPrivateInputs): WitnessMap {
```
Fixed

PhilWindle commented on yarn-project/bb-prover/src/prover/bb_prover.ts:
```diff
@@ -236,103 +252,98 @@ export class BBNativeRollupProver implements ServerCircuitProver {
+      const vkData = await this.updateVerificationKeyAfterProof(provingResult.vkPath!, circuitType);
+
+      // Read the proof and then cleanup up our temporary directory
+      const rawProof = await fs.readFile(`${provingResult.proofPath!}/${PROOF_FILENAME}`);
```
Oh, I think your are right actually. Well spotted. I hadn't realised they had converged to be so similar.

PhilWindle commented on yarn-project/end-to-end/src/e2e_prover/e2e_prover_test.ts:
```diff
@@ -120,49 +130,87 @@ export class ClientProverTest {
+      // stop the fake provers
+      proverAgents: 0,
+      realProofs: true,
+      minTxsPerBlock: 1, // min 2 txs per block
```
No, the comment is corrct, the code is wrong. Nice catch!

PhilWindle committed:
Review changes

PhilWindle committed:
Formatting

PhilWindle committed:
Try hardware concurrency

PhilWindle committed:
Conditional env sent to bb

PhilWindle committed:
Formatting

PhilWindle committed:
Removed debug log

PhilWindle committed:
Use verifiable trait

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

alexghr committed:
feat: verify block proofs on chain (#6568)

This PR adds commands to generate the verification Solidity contract for
the RootRollup so that block proofs can be validated onchain

Stacked on top of #6425

PhilWindle committed:
Merge branch 'master' into pw/more-recursion

PhilWindle committed:
Fix

PhilWindle committed:
Test fixes

PhilWindle committed:
Use verify trait
